packages/react-dom/src/client/ReactDOMInput.js

```diff
@@ -177,13 +177,22 @@ export function updateWrapper(element: Element, props: Object) {
 
   if (value != null) {
     if (type === 'number') {
-      if (
-        (value === 0 && node.value === '') ||
-        // We explicitly want to coerce to number here if possible.
-        // eslint-disable-next-line
-        node.value != (value: any)
-      ) {
-        node.value = toString(value);
+      let nodeValue = node.value;
+
+      // If the React prop value is NaN and current input value is invalid,
+      // it means both React and the DOM have invalid states (reported as
+      // empty). In this case, do not assign the value property, which would
+      // clear user input. The resulting state remains the same while still
+      // maintaining user expectations.
+      if (!(isNaN(value) && nodeValue === '')) {
```

Comment:
This actually trips up on browser that don't support number inputs.

I don't think we should execute this code path if the browser doesn't support number inputs. We can handle that by just checking `input.type` in IE9, it reports "text" instead of "number".

Reply:
Is IE9 the only browser we support that doesn't support number inputs? I feel like we're getting close to the point where we should consider dropping IE9 support.

Reply:
Ya. IE9 is the only one. I think we're getting close too - should we consider it?

Reply:
Looking at https://caniuse.com/usage-table: How come IE9 is still at 0.2%? 🙈

Reply:
It's been more than a year, IE9 is still 0.22% 👀